### PR TITLE
Build tools: remove all old files when cleaning

### PR DESCRIPTION
**Pull request by azaozz**
Based on https://github.com/WordPress/wordpress-develop/pull/3843.

Trac ticket: https://core.trac.wordpress.org/ticket/47749

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -110,27 +143,39 @@ module.exports = function(grunt) {
...
-			],
+
+			// Always clean built JS and CSS files from /src. Also from /build when building there.
+			// The build:css, build:js, etc. tasks may be run independently,
+			// and should clean all old files to avoid errors.
+			css: cssFiles.map( function( file ) {
+				return setFilePath( SOURCE_DIR, file );
+			} ).concat(
+				WORKING_DIR === BUILD_DIR
+				? cssFiles.map( function( file ) {
```
Comment (mukeshpanchal27) [inline Gruntfile.js]: @azaozz JavaScript coding standards shows error `Misleading line break before '?'; readers may interpret this as an expression boundary.`

Comment (azaozz) [reply Gruntfile.js]: Yea, this was formatted like Prettier (as in Gutenberg) but seems there are some "double standards" for JS (was planning to look more there but got distracted) :) 
Will try to work around the inconsistency.

Comment (azaozz) [reply Gruntfile.js]: Apparently no workaround... Prettier ternaries not allowed :)

Comment (mukeshpanchal27) [reply Gruntfile.js]: Awesome. Thanks for updating

File: Gruntfile.js [hunk 2]
```diff
@@ -32,6 +32,39 @@ module.exports = function(grunt) {
...
+			'wp-includes/css/*-rtl*.css',
+			'wp-admin/css/colors/**/*.css',
+			'wp-includes/blocks/**/*.css',
+		],
+
+		// All built js files, in /src or /build.
+		jsFiles = [
+			'wp-admin/js/',
+			'wp-includes/js/',
+			'wp-includes/blocks/**/*.js',
```
Comment (ironprogrammer) [inline Gruntfile.js:50]: There are a few `*.js.map` files that could also be cleaned up when running `npm run build:dev`:
```suggestion
			'wp-includes/blocks/**/*.js',
			'wp-includes/blocks/**/*.js.map',
```
What do you think?

**Review by ironprogrammer — COMMENTED** [review]
Hi, @azaozz 👋🏻 There were a few vestigial files when I ran the PR locally.

After `npm run build:dev`, there were:
```sh
$ git ls-files -o src/wp-includes
src/wp-includes/blocks/file/view.js.map
src/wp-includes/blocks/navigation/view-modal.js.map
src/wp-includes/blocks/navigation/view.js.map
```
so I've added a small suggestion for `jsFiles`.

After `npm run build`:
```sh
$ git ls-files -o src/wp-includes
src/wp-includes/js/tinymce/tinymce.js
```
For this one, during `vendor-js` [the file is copied specifically into `SOURCE_DIR` after `clean:js` has run](https://github.com/WordPress/wordpress-develop/blob/9132dc2b34840f6a1923e866986c1b777442426e/Gruntfile.js#L284). Should this file also have a dedicated clean up?

**Comment by azaozz** [comment]
Frankly I'm still "on the fence" between this approach and the approach from https://github.com/WordPress/wordpress-develop/pull/3843. 

In theory best would be to remove all "built" files from `/src` before bulk-copying to `/build`, or the built files should be excluded. In addition to keep the cleanup working the same as now, the js and css built files should be removed from the `WORKING_DIR` when building. Not sure how critical is that but probably "nice to keep".

Then perhaps the right approach would be to extend `clean:files` to clean everything in /build (as it does now) plus all build files from /src. This should ensure every new WP build starts with a clean slate.

**Comment by ironprogrammer** [comment]
Thanks, @azaozz! After the [`clean:files` update](https://github.com/WordPress/wordpress-develop/pull/4038/commits/d98f14077aab03f1b91e6978f9018fa4588e0736), there still remains the untracked file `src/wp-includes/js/tinymce/tinymce.js`. It sneaks past `clean` because it is copied to `src/` later, during `vendor:js`.

This file is somewhat of an exception in the `vendor:js` task. The original patch, [47749.1.diff](https://core.trac.wordpress.org/attachment/ticket/47749/47749.1.diff) (line 1408), handled this case by calling `clean:js` again at the end of `build` (the same was carried over to [PR 3843](https://github.com/WordPress/wordpress-develop/pull/3843/files#diff-8e7e5f941c67b0cf481a9d650295119d4922b96117e7adc069ba30a81a78c65fR1625)).

Should this be considered here as well?

**Comment by azaozz** [comment]
> there still remains the untracked file src/wp-includes/js/tinymce/tinymce.js

Good catch :) 
Seems cleaning of this was never fixed after moving all the js files to /src/js (in WP 5.0 I think). As `tinymce.js` is only needed in /src, seems it should be copied only when building there.

**Pull request closed without merging**